### DISCUSSION TITLE
Problems implementing command that needs a parser instance

### DISCUSSION THREAD

**Discussion by hallvard** [eclipse-langium/langium#141] [Q&A] [answered]
I'm trying to implement a generator from my language to html, and have created a Generator interface and a corresponding implementation that is included in the language's added services. My generator is invoked by a command (extension), first it parses the text in the active editor, then it calls the generator's generate method and finally the resulting html is written to a file.

The command is triggered correctly and the generator itself works, the problem is parsing the editor's document. I've defined my generator so it gets the parser injected, but in that process the GrammarAccess tries to load the grammar.json file from a path computed in the constructor as follows: `path.join(__dirname, 'grammar.json')`. However, `__dirname` is the **out** folder, not **out/language-server** where **grammar.json** resides, so it crashes during injection. If I add a segment to the join arguments, my code works, but the language server crashes. It seems that when the language server creates the GrammarAccess **__dirname** is set correctly, but not when my extension tries to create it (during injection). How am I supposed to do this?

Another issue I discovered when trying to auto-generate when a language file changes, is that the creation of the file watcher uses a faulty glob. In my case the call was `workspace.createFileSystemWatcher('**/*.[".quiz"]')` rather than `workspace.createFileSystemWatcher('**/*.quiz')`(which works). The value **[".quiz"]** seems to match value of the **extensions** field in package.json, which is a list, not a single extension (without the .) which it would need to be to work.

**Comment by hallvard** [comment]
My guess is that `__dirname` is set to the folder where the "main" entry point is, and the extension (`extension.ts`) gets a different one that the language server (`main.ts`). Hence, we'll need two grammar.json files.

In the domainmodel example I found the following in the `scripts` section in `package.json`:

```
     "langium:generate": "langium generate && npm run copygrammar",
      "copygrammar": "copyfiles -f ./src/language-server/generated/grammar.json ./out"
```

Also needed a devDependency: `"copyfiles": "^2.4.1"`

This creates a copy of the grammar.json where it will be found during injection in the extension, when you regenerate after making a grammar change.

This should work out-of-the-box in my opinion.

**Reply by msujew** [reply]
I wouldn't recomment using the `copyfiles` dev dependecy, since you can already use webpack's `CopyPlugin`, which is more of an all-in-one solution that our `copyfiles` call. That is done only internally in our own stuff, as we usually don't use webpack there for multiple reasons.

**Reply by msujew** [reply]
We actually didn't expect users to reference the `langium` dependency in their extension code at all. That's why we set the webpack build up as we did. We will discuss internally, whether we want to support such a use case out-of-the-box.

**Reply by hallvard** [reply]
My use case is generating a derived file triggered by a DSL file changed (or explicitly by means of a command). That's a pretty common thing to do, so should be supported, IMO. I'm used to Xtext and its injection mechanism (guice) so this (getting the injected Services instance) seems a natural solution to me. With the copying of the grammar.json file, it works very well! I'll create a public git repo with the code pretty soon. The code is a port of a small Xtext-based example I use to try languim. The goal is to offer it as an alternative to Xtext for a DSL class assignment.

**Reply by msujew** [reply]
We actually just discussed this now. This will be supported by PR #142. The relevant section is in [this file](https://github.com/langium/langium/blob/158c711947820eee8b94608540947231a6652806/examples/arithmetics/src/language-server/generated/grammar.ts) which replaces the `grammar.json` by a file which is easily picked up by webpack (or other bundling libraries). Therefore, eliminating the need for the whole copying process/build step.

It's really cool that you are planning to use Langium in your classes. Feel free to ask any further questions on Langium, I'm always happy to answer/help with anything that comes up.

**Reply by hallvard** [reply]
Great! I expect students to be more motivated when they can create a VSCode extension rather than an Eclipse plugin.
BTW, here's the repo: https://github.com/hallvard/langium-quiz

**Comment by msujew** [comment, marked as answer]
Hi @hallvard, glad to hear you are using Langium.

Regarding your first issue: The easiest way would probably just to add another call to the webpack `CopyPlugin` to add another copy of the `grammar.json` to your output directory. You would add something like the following to the `vscodeConfig` and point the `to` property to the `out` directory.

https://github.com/langium/langium/blob/ba581331bd65733ff1f74a5950320fbbad523ce3/packages/generator-langium/langium-template/webpack.config.js#L49-L58

An alternative would be to direct the `main.js` bundle with the `grammar.json` to the `out` directory instead of `out/language-server`. Beware that you have to remove one `../` from the `devtoolModuleFilenameTemplate` of the `lspConfig` or debugging of the language server will stop working. You also need to change the module path in the `extension.ts`.

A more involved alternative would be to call the `executeCommand` endpoint of the language server, instead of letting the vscode extension do the work. But that's quite hard to do as long as we don't add a service for that to Langium.

Thanks for the other issue, that seems to be a bug in our yeoman generator.
